src/main.js

```diff
@@ -16,6 +16,28 @@ function renderWidget() {
         sidebarEl.insertBefore(widgetRootEl, sidebarEl.firstChild);
 
         ReactDOM.render(<Widget />, widgetRootEl);
+        var rectangle = widgetRootEl.getBoundingClientRect();
+        var children = widgetRootEl.parentElement.children;
+        var maxDistance = 0;
+        for (let i = 0; i < children.length; i++){
+            if (children[i] != widgetRootEl){
+                let otherRectangle = children[i].getBoundingClientRect();
+                let num = rectangle.bottom - otherRectangle.top;
+                if (otherRectangle.height > 0 && num > maxDistance){
+                    maxDistance = num;
+                }
+            }
+        }
+        for (let i = 0; i < children.length; i++){
+            if (children[i] != widgetRootEl){
+                let otherRectangle = children[i].getBoundingClientRect();
+                children[i].style.marginTop = otherRectangle.top.toString() + "px";
+                let newOtherRectangle = children[i].getBoundingClientRect();
+                let previousMarginDistance = newOtherRectangle.top - otherRectangle.top;
+                let num = otherRectangle.top - previousMarginDistance + maxDistance;
+                children[i].style.marginTop = num.toString() + "px";
+            }
+        }
```

Comment:
Hmm, this is really interesting.

The code is quite complicated though and I think there could be a cleaner solution. Also, the *Replace video* link gets the margin too (which is something we probably don't want). 
I came up with two possible solutions. The video element uses negative margin-top so I thought that we could just make the margin-top zero. Like this:

```js
        var videoPreviewEl = sidebarEl.querySelector('.video-preview');
        if (videoPreviewEl) {
            videoPreviewEl.style.marginTop = 0;
        }
```

We can generalize it to all element with negative margin-top in the sidebar which is probably better as Last.fm can add another similar element in the future.

```js
        var children = sidebarEl.children;
        for (let i = 0; i < children.length; i++){
            let child = children[i];
            let style = window.getComputedStyle(child);
            if (parseInt(style.marginTop) < 0) {
                child.style.marginTop = 0;
            }
        }
```

Another option could be shifting the Friend Who Listen widget below original widget in top section of the sidebar (like the video or Similar To) but I don't know if people don't actually like the prominent position of Friends Who Listen widget.

I also discovered that album arts don't work with neither my or your approach. They don't use negative margins but something different and I wasn't able to find out what. For example [here](https://www.last.fm/music/Arcade+Fire/Reflektor).
<img width="411" alt="Screenshot 2020-12-28 at 23 13 25" src="https://user-images.githubusercontent.com/3393339/103246025-4baa3280-4962-11eb-9e4d-536a0a8d062c.png">

What do you think about it? Which approach do you like more? And do you have any ideas what to do with the album arts?

Reply:
Ah, so I found out that the album art element which has negative margins is not one of sidebarEl's children, but instead one of its _children's children_, with its class as `album-overview-cover-art`.

I do like your generalized code approach a lot, although for it to work with the album art, we probably need to check marginTop of not only sidebarEl's children but also its grandchildren, which might lead to more complicated code. I think your video code approach could be modified into a brief manner of checking both video preview and album art elements. For instance, I tried this, and it seems to work:
```javascript
var videoOrArtEl = sidebarEl.querySelector('.video-preview') || sidebarEl.querySelector('.album-overview-cover-art');
if (videoOrArtEl) {
      videoOrArtEl.style.marginTop = 0;
}
```
I think I would choose to do an approach like this, though I realize that this could very well need to be updated if Last.fm goes through another overhaul in the future, so let me know if you want to me to add a commit with this solution or another one. By the way, regarding the position of the widget, I personally do like its prominent position near the top.

Reply:
Oh, you're right! I was inspecting only the direct children and I didn't realise that that their descendants can have negative margins too.
I agree the code would get too complicated if we had to check all nested descendants and I really like your solution.

I though about a bit more general approach that would work in case there were more elements with negative margins in the future but that probably won't happen as it doesn't make much sense.
```js
var negativeMarginSelectors = ['.video-preview', '.album-overview-cover-art'];
negativeMarginSelectors.forEach(selector => {
    var element = sidebarEl.querySelector(selector);
    if (element) {
        element.style.marginTop = 0;
    }
});
```

So I am adding the snippet here only for record and I am completely fine with your approach. Could you add a commit with this (videoOrArtEl) please?

I will merge it afterwards and release a new version.

Reply:
Alright, committed it.

Reply:
Thank you!